protocol/adapter/datasync_whisper.go

```diff
@@ -0,0 +1,173 @@
+package adapter
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"log"
+	"sort"
+
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/status-im/mvds/node"
+	"github.com/status-im/mvds/protobuf"
+	"github.com/status-im/mvds/state"
+	dstrns "github.com/status-im/mvds/transport"
+
+	dspeer "github.com/status-im/status-console-client/protocol/datasync/peer"
+	"github.com/status-im/status-console-client/protocol/subscription"
+	"github.com/status-im/status-console-client/protocol/transport"
+	"github.com/status-im/status-console-client/protocol/v1"
+
+	whisper "github.com/status-im/whisper/whisperv6"
+)
+
+type PacketHandler interface {
+	AddPacket(dstrns.Packet)
+}
+
+type DataSyncWhisperAdapter struct {
+	node      *node.Node
+	transport transport.WhisperTransport
+	packets   PacketHandler
+}
+
+// DataSyncWhisperAdapter must implement Protocol interface.
+var _ protocol.Protocol = (*DataSyncWhisperAdapter)(nil)
+
+func NewDataSyncWhisperAdapter(n *node.Node, t transport.WhisperTransport, h PacketHandler) *DataSyncWhisperAdapter {
+	return &DataSyncWhisperAdapter{
+		node:      n,
+		transport: t,
+		packets:   h,
+	}
+}
+
+// Subscribe listens to new messages.
+func (w *DataSyncWhisperAdapter) Subscribe(
+	ctx context.Context,
+	messages chan<- *protocol.Message,
+	options protocol.SubscribeOptions,
+) (*subscription.Subscription, error) {
+	if err := options.Validate(); err != nil {
+		return nil, err
+	}
+
+	filter := newFilter(w.transport.KeysManager())
+	if err := updateFilterFromSubscribeOptions(filter, options); err != nil {
+		return nil, err
+	}
+
+	// Messages income in batches and hence a buffered channel is used.
+	in := make(chan *whisper.ReceivedMessage, 1024)
+	sub, err := w.transport.Subscribe(ctx, in, filter.Filter)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for item := range in {
+			payload, err := w.decodePayload(item)
+			if err != nil {
+				log.Printf("failed to decode message %#+x: %v", item.EnvelopeHash.Bytes(), err)
+				continue
+			}
+
+			packet := dstrns.Packet{
+				Group:   toGroupId(item.Topic),
+				Sender:  dspeer.PublicKeyToPeerID(*item.Src),
+				Payload: payload,
+			}
+			w.packets.AddPacket(packet)
+
+			for _, m := range w.decodeMessages(payload) {
+				m.SigPubKey = item.Src
+				messages <- m
+			}
+		}
+	}()
+
+	return sub, nil
+}
+
+func (w *DataSyncWhisperAdapter) decodePayload(message *whisper.ReceivedMessage) (payload protobuf.Payload, err error) {
```

Comment:
I guess this is fine, I am not usually a fan of named return variables though

Reply:
Me neither but in this case it works great :)